src/bin/miri.rs

```diff
@@ -9,22 +10,43 @@ extern crate rustc_errors;
 
 use std::convert::TryFrom;
 use std::env;
+use std::path::PathBuf;
+use std::rc::Rc;
 use std::str::FromStr;
 
 use hex::FromHexError;
 use log::debug;
 
-use rustc_session::{CtfeBacktrace, config::ErrorOutputType};
+use rustc_session::{CtfeBacktrace, config::ErrorOutputType, search_paths::PathKind};
 use rustc_errors::emitter::{HumanReadableErrorType, ColorConfig};
 use rustc_driver::Compilation;
 use rustc_hir::def_id::LOCAL_CRATE;
-use rustc_middle::ty::TyCtxt;
+use rustc_interface::interface::Config;
+use rustc_middle::{
+    ty::{query::Providers, TyCtxt},
+    middle::exported_symbols::{ExportedSymbol, SymbolExportLevel},
+};
 
 struct MiriCompilerCalls {
     miri_config: miri::MiriConfig,
 }
 
 impl rustc_driver::Callbacks for MiriCompilerCalls {
+    fn config(&mut self, config: &mut Config) {
+        config.override_queries = Some(|_, _, external_providers| {
+            external_providers.used_crate_source = |tcx, cnum| {
+                let mut providers = Providers::default();
+                rustc_metadata::provide_extern(&mut providers);
+                let mut crate_source = (providers.used_crate_source)(tcx, cnum);
+                // rustc will emit "crate ... required to be available in rlib format, but was not
+                // found in this form" errors once we use `tcx.dependency_formats()` if there's no
+                // rlib provided, so setting a dummy path here to workaround those errors.
+                Rc::make_mut(&mut crate_source).rlib = Some((PathBuf::new(), PathKind::All));
```

Comment:
Hm, okay, no idea how bad of a hack this one is.^^ Cc @oli-obk 
Please add a "HACK:" label.